src/main/java/com/commercetools/sync/services/impl/ProductServiceImpl.java

```diff
@@ -117,104 +87,61 @@ public ProductServiceImpl(@Nonnull final ProductSyncOptions syncOptions) {
-
-    if (keys.isEmpty()) {
-      return CompletableFuture.completedFuture(Collections.emptySet());
-    }
-
-    final List<List<String>> chunkedKeys = ChunkUtils.chunk(keys, CHUNK_SIZE);
-
-    final List<ProductProjectionQuery> keysQueryMapperList =
-        chunkedKeys.stream()
-            .map(
-                _keys ->
-                    keysQueryMapper
-                        .apply(new HashSet<>(_keys))
-                        .withLimit(CHUNK_SIZE)
-                        .withFetchTotal(false))
-            .collect(toList());
-    return ChunkUtils.executeChunks(syncOptions.getCtpClient(), keysQueryMapperList)
-        .thenApply(ChunkUtils::flattenPagedQueryResults)
-        .thenApply(
-            chunk -> {
-              chunk.forEach(
-                  resource -> keyToIdCache.put(keyMapper.apply(resource), resource.getId()));
-              return new HashSet<>(chunk);
-            });
-  }
-
   @Nonnull
   @Override
   public CompletionStage<Optional<ProductProjection>> fetchProduct(@Nullable final String key) {
 
-    if (isBlank(key)) {
-      return CompletableFuture.completedFuture(Optional.empty());
-    }
-
-    return syncOptions
-        .getCtpClient()
-        .execute(
+    return fetchResource(
+        key,
+        () ->
             ProductProjectionQuery.ofStaged()
-                .withPredicates(buildProductKeysQueryPredicate(singleton(key))))
-        .thenApply(
-            pagedQueryResult ->
-                pagedQueryResult
-                    .head()
-                    .map(
-                        resource -> {
-                          keyToIdCache.put(key, resource.getId());
-                          return resource;
-                        }));
+                .withPredicates(buildProductKeysQueryPredicate(singleton(key))));
   }
 
   @Nonnull
   @Override
   public CompletionStage<Optional<ProductProjection>> createProduct(
       @Nonnull final ProductDraft productDraft) {
+
     return createResource(productDraft, ProductCreateCommand::of)
         .thenApply(product -> product.map(opt -> opt.toProjection(STAGED)));
   }
 
   @Nonnull
+  @Override
   public CompletionStage<ProductProjection> updateProduct(
-      @Nonnull final ProductProjection product,
+      @Nonnull final ProductProjection productProjection,
       @Nonnull final List<UpdateAction<Product>> updateActions) {
 
-    return updateResourceByKey(product, updateActions);
+    return updateResourceAndMapToProjection(productProjection, updateActions);
   }
-  // TODO: (JVM-SDK), remove this replicated code, when the the JVM-SDK allows to use
-  // ProductProjections in a ProductUpdateCommand.
 
-  // see: https://github.com/commercetools/commercetools-sync-java/pull/699#discussion_r600457980
   @Nonnull
-  CompletionStage<ProductProjection> updateResourceByKey(
-      @Nonnull final ProductProjection resource,
+  CompletionStage<ProductProjection> updateResourceAndMapToProjection(
```

Comment:
Where in the code is this part `MapToProjection`?

Reply:
https://github.com/commercetools/commercetools-sync-java/pull/714/commits/849dd8aba172d4cd02a138e6dfd75635f3fa80c3
check `.thenApply(p -> p.toProjection(STAGED)));`